# <center>Stochastic Search Optimization Problem using Random Walk Concept with-in a Constrained Search Space

[UPdate Code from Previous Jupyter Notebook](https://github.com/Shayankr/Machine-Learning-with-Python/blob/main/Stochastic_Search_Random_Walk.ipynb) Here I have used more optimized concept with same constrained criteria.

## Here, I am generating "k" no. of points around each initial solution points (no_of_points) and then finally choosing -- that one point-- which is most optimum.

In [1]:
import random
import numpy as np

def objective_function(x):
    x1, x2 = x
    # Define your objective function here
    return -(5-(x1-2)**2-(x2-3)**2)

def random_walk(obj_func, population, number_of_generations, step_size, fitness_threshold, x_min, x_max,k):
    best_fitness = float("inf")
    best_solution = None
    
#     # Generate the initial solutions
#     initial_solutions = []
#     for i in range(k):
#         x1 = random.uniform(x_min[0], x_max[0])
#         x2 = random.uniform(x_min[1], x_max[1])
#         initial_solutions.append((x1, x2))
    
    # Evaluate the fitness of the initial population
    fitness_scores = np.array([obj_func(x) for x in population])
    fitness_scores = fitness_scores-min(fitness_scores)  # making fitness value non-negative
    fitness_scores = fitness_scores/sum(fitness_scores)   # Normalize the fitness value
    
    # Main loop
    for i in range(number_of_generations):
        # Update the best solution if a better one is found
        best_index = np.argmin(fitness_scores)
        if fitness_scores[best_index] < best_fitness:
            best_fitness = fitness_scores[best_index]
            best_solution = population[best_index]
        
        # Terminate the loop if the fitness threshold is reached
        if best_fitness <= fitness_threshold:
            break
        
        # Generate a new population by adding a random step to each solution
        new_population = []
        for x in population:
            x1=[]
            x2=[]
            for j in range(k):
                new_x1 = x[0] + random.uniform(-step_size, step_size)
                new_x2 = x[1] + random.uniform(-step_size, step_size)
                if ((x_min<=new_x1 and new_x1<=x_max) and (x_min<=new_x1<=x_max)):
                    x1.append(new_x1)
                    x2.append(new_x2)
                else:
                    continue
            # now find the best value among newly generated pointa/solution sets around a particular previous solution point
            for p,q in zip(x1,x2):
                if objective_function((p,q))<objective_function(x):
                    x=(p,q)
                else:
                    pass
            # Now, I got the best solution point around a previous point among newly generated points
            # now go for the next previous point and repeat the same procedure.
            
            # Ensure that the new solution is within the search space
#             new_x1 = max(x_min[0], min(x_max[0], new_x1))
#             new_x2 = max(x_min[1], min(x_max[1], new_x2))
            
#             new_population.append((new_x1, new_x2))
            new_population.append(x)
            

            
        population = new_population
        
        # Evaluate the fitness of the new population
        fitness_scores = np.array([obj_func(x) for x in population])
        fitness_scores = fitness_scores-min(fitness_scores)
        fitness_scores = fitness_scores/sum(fitness_scores)
    
    return best_solution

# Define the initial population
initial_population = [(-2, 2), (1, 2), (0, 0), (-1, -2), (3, 1)]

# Define the search space boundaries
x_min = (-3, -5)
x_max = (2, 4)

result = random_walk(objective_function, initial_population, number_of_generations=1000, step_size=0.4, fitness_threshold=1e-6, x_min=x_min, x_max=x_max, k=1000)
print("Best solution:", result)
print(f"The value of objective function at given initial popluation or solution sets are: {[objective_function(i) for i in initial_population]}")
print(f"The value of the objective function at best solution is: {objective_function(result)}")

Best solution: (1, 2)
The value of objective function at given initial popluation or solution sets are: [12, -3, 8, 29, 0]
The value of the objective function at best solution is: -3


In [2]:
# import random
# random.seed(seed)

In [3]:
random.uniform(0,1)

0.06235593491078084

In [4]:
list_=[1,0,5,4]
np.argmin(list_)

1

### Generating k random initial points instead of 

In [5]:
import random
import numpy as np

def objective_function(x):
    x1, x2 = x
    # Define your objective function here
    return -(5-(x1-2)**2-(x2-3)**2)

def random_walk(obj_func, number_of_generations, step_size, fitness_threshold, x_min, x_max,no_of_points,k,seed=None):
    random.seed(seed)
    
    best_fitness = float("inf")
    best_solution = None
    
    # Generate the initial solutions
    initial_solutions = []
    for i in range(no_of_points):
        x1 = random.uniform(x_min[0], x_max[0])
        x2 = random.uniform(x_min[1], x_max[1])
        initial_solutions.append((x1, x2))
    
    # Evaluate the fitness of the initial population
    fitness_scores = np.array([obj_func(x) for x in initial_solutions])
    fitness_scores = fitness_scores-min(fitness_scores)  # making fitness value non-negative
    fitness_scores = fitness_scores/sum(fitness_scores)   # Normalize the fitness value
    
    # Main loop
    for i in range(number_of_generations):
        # Update the best solution if a better one is found
        best_index = np.argmin(fitness_scores)
        if fitness_scores[best_index] < best_fitness:
            best_fitness = fitness_scores[best_index]
            best_solution = initial_solutions[best_index]
        
        # Terminate the loop if the fitness threshold is reached
        if best_fitness <= fitness_threshold:
            break
        
        # Generate a new population by adding a random step to each solution
        new_population = []
        for x in initial_solutions:
            x1=[]
            x2=[]
            for j in range(k):
                new_x1 = x[0] + random.uniform(-step_size, step_size)
                new_x2 = x[1] + random.uniform(-step_size, step_size)
                if ((x_min<=new_x1 and new_x1<=x_max) and (x_min<=new_x1<=x_max)):
                    x1.append(new_x1)
                    x2.append(new_x2)
                else:
                    continue
            # now find the best value among newly generated pointa/solution sets around a particular previous solution point
            for p,q in zip(x1,x2):
                if objective_function((p,q))<objective_function(x):
                    x=(p,q)
                else:
                    pass
            # Now, I got the best solution point around a previous point among newly generated points
            # now go for the next previous point and repeat the same procedure.
            
            # Ensure that the new solution is within the search space
#             new_x1 = max(x_min[0], min(x_max[0], new_x1))
#             new_x2 = max(x_min[1], min(x_max[1], new_x2))
            
#             new_population.append((new_x1, new_x2))
            new_population.append(x)

            
        initial_solutions = new_population
        
        # Evaluate the fitness of the new population
        fitness_scores = np.array([obj_func(x) for x in population])
        fitness_scores = fitness_scores-min(fitness_scores)
        fitness_scores = fitness_scores/sum(fitness_scores)
    
    return best_solution

# # Define the initial population
# initial_population = [(-2, 2), (1, 2), (0, 0), (-1, -2), (3, 1)]

# Define the search space boundaries
x_min = (-3, -5)
x_max = (2, 4)

result = random_walk(objective_function, number_of_generations=1000, step_size=0.2, fitness_threshold=1e-6, x_min=x_min, x_max=x_max,no_of_points=80, k=79, seed=123)
print("Best solution:", result)
# print(f"The value of objective function at given initial popluation or solution sets are: {[objective_function(i) for i in initial_population]}")
print(f"The value of the objective function at best solution is: {objective_function(result)}")

Best solution: (1.9763250682900022, 3.0172783147874664)
The value of the objective function at best solution is: -4.999140957446632


#### In above code:

### By choosing 80 initial points and 79 points around each initial points then selecting 1(one) best among 79 points to consider into next generations....achieve best solution: [-4.999140957446632]